avocado/core/safeloader.py

```diff
@@ -30,38 +31,57 @@ class AvocadoModule(object):
     """
     Representation of a module that might contain avocado.Test tests
     """
-    __slots__ = 'path', 'test_import', 'mod_import', 'mod'
+    __slots__ = 'path', 'test_import', 'mod_import', 'mod', 'imported_objects'
 
     def __init__(self, path, test_import=False, mod_import=False):
         self.path = path
         self.test_import = test_import
         self.mod_import = mod_import
+        # Dict where:
+        #   key => object name how is visible from this module
+        #   value => Something-like a directory path to the import.
+        #            Basicall it's $path/$module/$variable, but depending
+        #            on import could be also $path/$module.
+        self.imported_objects = {}
         if os.path.isdir(path):
-            self.path = os.path.join(path, "__init__.py")
-        else:
-            self.path = path
-        with open(self.path) as source_file:
-            self.mod = ast.parse(source_file.read(), self.path)
+            path = os.path.join(path, "__init__.py")
+        self.path = path
+        with open(path) as source_file:
+            self.mod = ast.parse(source_file.read(), path)
+
+    def add_imported_object(self, statement):
+        path = os.path.abspath(os.path.dirname(self.path))
+        if hasattr(statement, 'module'):
+            module_path = statement.module.replace('.', os.path.sep)
+            path = os.path.join(path, module_path)
+        for name in statement.names:
+            path = os.path.join(path, name.name.replace('.', os.path.sep))
+            if name.asname is None:
+                self.imported_objects[name.name] = path
+            else:
+                self.imported_objects[name.asname] = path
 
     def iter_classes(self):
         """
         Iter through classes and keep track of imported avocado statements
         """
         for statement in self.mod.body:
             # Looking for a 'from avocado import Test'
-            if (isinstance(statement, ast.ImportFrom) and
-                    statement.module == 'avocado'):
-
-                for name in statement.names:
-                    if name.name == 'Test':
-                        if name.asname is not None:
-                            self.test_import = name.asname
-                        else:
-                            self.test_import = name.name
-                        break
+            if isinstance(statement, ast.ImportFrom):
+                self.add_imported_object(statement)
```

Comment:
It looks like `add_import_object` loops the statements the same way as its done two lines below.  It seems like a good candidate for breaking into a reusable function.

Reply:
You're right, it may add complexity to the code.